vllm/hpu/ops.py

```diff
@@ -25,72 +25,247 @@ def gelu_new(output, input):
+
+    # Similarly to block mask, we'll create s 2D tensor of token indices per each block,
+    # which we'll then transform into a boolean tensor with mask
+    seq_indices = torch.arange(block_size, dtype=torch.int64, device=device).view(1, -1)
+    seq_indices = seq_indices.expand(num_seqs, seq_indices.size(1))
+
+    # Create mask with 1s for all tokens that are fully out of bound, and 0s for the rest.
+    # We apply a bias of block_size for sequences that have context length divisible by block_size,
+    # as we don't want to clear anything within their last block - it is fully filled
+    last_block_offsets = (context_lens % block_size + block_size * (context_lens % block_size == 0)).view(-1, 1)
+    seq_mask = seq_indices >= last_block_offsets
+
+    # Apply block mask to weights to diagonal (num_seq, num_query_heads, 1, block_size) tensor.
+    last_blocks_diag.masked_fill_(seq_mask.view(num_seqs, 1, 1, block_size), attn_weights_blocks_filler)
+
+    # Scatter the (num_seq, num_query_heads, 1, block_size) tensor back into attn_weights_blocks using
+    # the same indices as we did in gathering. Each "row" will be stored at (last_block_index[i],i),
+    # where i is sequence index.
+    seq_idx = torch.arange(num_seqs, dtype=torch.int64, device=device)
+    edge_indices = (last_block_indices, seq_idx)
+    attn_weights_blocks.index_put_(edge_indices, last_blocks_diag)
+
+    # NOTE(kzawora): For whatever reason, we also need to cleanup values, otherwise vLLM inference will return garbage.
+    # I'll need to figure out why. For now, we'll follow exactly the same steps as we did with weight cleanup.
+    if sanitize_values:
+        # For filling fully OoB value blocks, we can use the same mask as we did previously.
+        value_blocks.masked_fill_(attn_weights_blocks_mask, value_blocks_filler)
+
+        # Same thing for filling partial blocks - obtain all last blocks for all sequences
+        last_values = value_blocks.index_select(0, last_block_indices)  # [num_seqs, num_seqs, num_kv_heads, block_size,  head_size]
+
+        # Reduce the tensor to have only a single last block per each sequence
+        last_values_diag = torch.diagonal(last_values, dim1=0, dim2=1, offset=0).permute((3, 0, 1, 2))
+
+        # Mask blocks
+        last_values_diag.masked_fill_(seq_mask.view(num_seqs, 1, block_size, 1), value_blocks_filler)
+
+        # Put masked blocks back in their place
+        value_blocks.index_put_(edge_indices, last_values_diag)
+
+    # Phew, that was a lot.
+
+def paged_attention_v1(
+    query_in,
+    key_cache_in,
+    value_cache_in,
+    head_mapping,
+    scale,
+    block_tables,
+    context_lens,
+    block_size,
+    max_context_len,
+    alibi_slopes,
+    attn_masks=None,
+) -> None:
+    sanitize_values = True
+    device = query_in.device
+    query = query_in
+    key_cache = key_cache_in
+    value_cache = value_cache_in
```

Comment:
The `*_in` args were used in a prior version to facilitate type casting, since FP16 was causing precision issues in softmax. If that is no longer the case, then the arg names should be changed and these lines should be removed.